lib/environment/permissions.js

```diff
@@ -9,52 +9,44 @@ function filterPerms(perms) {
 	return { permissions, origins };
 }
 
-const STORAGE_KEY = 'RES.cachedDeniedPerms';
+export const has = _.memoize(
+	async (...perms) => {
+		const { permissions, origins } = filterPerms(perms);
 
-export function reset() {
-	Storage.delete(STORAGE_KEY);
-}
+		return await sendMessage('permissions', { operation: 'contains', permissions, origins });
+	},
+	(...perms) => perms.join(',')
+);
 
 /**
+ * @param {string} perms Name of site to request permission for
  * @param {...string} perms Optional Chrome permissions to request.
  * @returns {Promise<void, *>} Resolves if the permissions are granted, rejects otherwise.
  */
-export const request = _.memoize(
-	mutex(async (...perms) => {
-		const { permissions, origins } = filterPerms(perms);
-
-		const hasPermissions = await sendMessage('permissions', { operation: 'contains', permissions, origins });
-
-		// already granted, don't ask the user again
-		if (hasPermissions) return;
+export const request = mutex(async (siteName, ...perms) => {
+	if (await has(...perms)) return;
 
-		const cacheKey = perms.join(',');
-		const cachedPerms = await Storage.get(STORAGE_KEY) || {};
-		if (cachedPerms[cacheKey]) {
-			throw new Error(`Cached denial for permission: ${perms.join(', ')}`);
-		}
-
-		const urlStripRe = /((?:\w+\.)+\w+)(?=\/|$)/i;
-
-		await Alert.open(`
-			<p>To create expandos, RES needs permission to access the API(s) at:</p>
-			<br>
-			<p>${origins.map(url => `<code>${urlStripRe.exec(url)[0]}</code>`).join(', ')}</p>
-			<br>
-			<p>Be assured RES does not access/modify any of your information on these domains - it only accesses the public API.</p>
-			<br>
-			<p><b>Clicking "ok" will open a popup where you may accept or deny this permission.</b></p>
-		`);
-
-		const granted = await sendMessage('permissions', { operation: 'request', permissions, origins });
+	const { permissions, origins } = filterPerms(perms);
 
-		if (!granted) {
-			Storage.patch(STORAGE_KEY, { [cacheKey]: true });
-			throw new Error(`Permission not granted for: ${perms.join(', ')}`);
-		}
-	}),
-	(...perms) => perms.join(',')
-);
+	const urlStripRe = /((?:\w+\.)+\w+)(?=\/|$)/i;
+
+	Notifications.showNotification({
+		header: 'Permission required',
+		moduleID: 'permissions',
+		closeDelay: 20000,
+		message: `
+				<p>In order to inline expand content from ${siteName}, RES needs permission to access the API(s) at:</p>
```

Comment:
A lot of users won't know what an "API" is - What about just saying "RES needs permission to access these sites"?